agent/agentpb/encoder.go

```diff
@@ -0,0 +1,35 @@
+package agentpb
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+type ProtoMarshaller interface {
+	Size() int
+	MarshalTo([]byte) (int, error)
+	Unmarshal([]byte) error
+	ProtoMessage()
+}
+
+func EncodeInterface(t structs.MessageType, message interface{}) ([]byte, error) {
+	if marshaller, ok := message.(ProtoMarshaller); ok {
+		return Encode(t, marshaller)
+	}
+	return nil, fmt.Errorf("message does not implement the ProtoMarshaller interface")
```

Comment:
Might be useful to slip `%T` in this message so you have a hope of knowing what type it was

Reply:
I actually considered panic'ing here at first too as it really is a fundamental problem similar to not being able to process unknown raft request types. I decided against that so I will add the %T